generators/beagle/src/main/scala/BeaglePeriphery.scala

```diff
@@ -22,85 +22,157 @@ case object PeripheryBeagleKey extends Field[BeagleParams]
+{
   implicit val p: Parameters
 
   val boot          = Input(Bool())
   val rst_async     = Input(Bool())
 
-  val clk_sel       = Output(UInt(p(PeripheryBeagleKey).clkSelBits.W))
-  val lbwif_divider = Output(UInt(p(PeripheryBeagleKey).lbwifDividerBits.W))
-  val hbwif_rsts    = Output(Vec(p(HbwifNumLanes), Bool()))
+  val uncore_clk_divisor = Output(UInt(p(PeripheryBeagleKey).uncoreClkDivBits.W))
+  val bh_clk_divisor     = Output(UInt(p(PeripheryBeagleKey).bhClkDivBits.W))
+  val rs_clk_divisor     = Output(UInt(p(PeripheryBeagleKey).rsClkDivBits.W))
+  val bh_clk_out_divisor = Output(UInt(p(PeripheryBeagleKey).bhClkDivBits.W))
+  val rs_clk_out_divisor = Output(UInt(p(PeripheryBeagleKey).rsClkDivBits.W))
+  val lbwif_clk_divisor  = Output(UInt(p(PeripheryBeagleKey).lbwifClkDivBits.W))
 
-  val switcher_sel  = Output(Bool())
+  val uncore_clk_pass_sel = Output(Bool()) // used to choose the undivided uncore clock or divided
+
+  val hbwif_rsts   = Output(Vec(p(HbwifNumLanes), Bool()))
+  val switcher_sel = Output(Bool())
 }
 
-trait HasBeagleTopModuleContents extends MultiIOModule with HasRegMap {
+/**
+ * MMIO controlled components
+ */
+trait HasBeagleTopModuleContents extends MultiIOModule with HasRegMap
+{
   val io: HasBeagleTopBundleContents
   implicit val p: Parameters
   def params: BeagleParams
   def c = params
 
-  val clk_sel = withReset(io.rst_async) {
-    Module(new AsyncResetRegVec(w = c.clkSelBits, init = 0))
+  // main clock divider regs
+  val r_uncore_clk_divisor = withReset(io.rst_async) {
+    Module(new AsyncResetRegVec(w = c.uncoreClkDivBits, init = c.uncoreClkDivInit))
+  }
+
+  val r_bh_clk_divisor = withReset(io.rst_async) {
+    Module(new AsyncResetRegVec(w = c.bhClkDivBits, init = c.bhClkDivInit))
+  }
+
+  val r_rs_clk_divisor = withReset(io.rst_async) {
+    Module(new AsyncResetRegVec(w = c.rsClkDivBits, init = c.rsClkDivInit))
+  }
+
+  // output clock divider regs
+  val r_bh_clk_out_divisor = withReset(io.rst_async) {
+    Module(new AsyncResetRegVec(w = c.bhClkDivBits, init = c.bhClkDivInit))
+  }
+
+  val r_rs_clk_out_divisor = withReset(io.rst_async) {
+    Module(new AsyncResetRegVec(w = c.rsClkDivBits, init = c.rsClkDivInit))
+  }
+
+  // lbwif clock divider reg
+  val r_lbwif_clk_divisor = withReset(io.rst_async) {
+    Module(new AsyncResetRegVec(w = c.lbwifClkDivBits, init = c.lbwifClkDivInit))
   }
-  val lbwif_divider = withReset(io.rst_async) {
-    Module(new AsyncResetRegVec(w = c.lbwifDividerBits, init = p(LbwifDividerInit)))
+
+  // uncore select bit reg
+  val r_uncore_clk_pass_sel = withReset(io.rst_async) {
+    Module(new AsyncResetRegVec(w = 1, init = 0))
   }
-  val hbwif_rsts = Seq.fill(p(HbwifNumLanes)) {
+
+  // hbwif reset signals
+  val r_hbwif_rsts = Seq.fill(p(HbwifNumLanes)) {
     withReset(io.rst_async) {
       Module(new AsyncResetRegVec(w = 1, init = 1))
     }
   }
-  val switcher_sel = withReset(io.rst_async) {
+
+  // connect hbwif or lbwif to memory system
```

Comment:
Should consider whether you want to have each hbwif lane individually useable (in case only 1 of them works)